File: Matrix.py
#Reimagining this project with a Matrix class object
#It has a size (mxn), a state (which version/step of the matrix it is), and entries
#It has the three basic row operations on it: swap, add, scale.
#eventually these matrices could be multiplied, etc. But that's another project

from math import gcd

#helper function 
#note: math.gcd only accepts integers
def reduce_frac (x:int,y:int):
    g_c_d = gcd(x,y)
    x = x // g_c_d
    y = y // g_c_d
    return ((x,y))

def fraction_multiplier(no_1: tuple,no_2: tuple) -> tuple: #multiplies two tuples with numerator in position 0 and denominator in pos. 1
    numerator = no_1[0] * no_2[0]
    denominator = no_1[1] * no_2[1]
    solution = reduce_frac(numerator, denominator)
    return solution

class Matrix:
    """A matrix has a size mxn, state (which version/step it is in the reduction process) and entries represented as tuples, 
    with numerator as the first entry and denominator as the second.
    Entries are a list with a list of tuples for each row.
    
    Usage example:
        matrix1 = Matrix((3,2), (0,1), [[(2,1),(4,1)],[(0,1),(-7,1)],[(3,1),(1,1)]])
        matrix2 = Matrix((3,3)), (2,3), [[(1,1),(1,2),(1,10)],[(0,1),(-7,1),(-4,1)],[(3,1),(1,1),(-1,1)]])
        
    """
    def __init__(self, size: tuple, state: tuple, entries: list):
        """A matrix with a size m x n, and entries 
        The size m x n is a tuple in order (m,n).
        The state is in the form (version, step). A new version begins after change following a backtrack
        """
        assert len(entries) == size[0], f"number of rows must match m of m x n matrix size"
        assert [len(i) == size[1] for i in entries], f"number of columns must match n in m x n matrix size"
        self.size = size
        self.state = state 
        self.entries = entries



    def __str__(self) -> str:
        """The textual format of a matrix is
        a_{11} a_{12} ... a_{1n}
        a_{21} a_{22} ... a_{2n}
        ...
        a_{m1} a_{m2} ... a_{mn}
        
        """
        return "\n".join(str(self.entries[i]) for i in range(self.size[0]))


    def __repr__(self) -> str:
        return f"(({self.size}, {self.state},{self.entries}))"
    

    #operator: 0 = division, 1 = multiplication. operand: (numerator,denominator)
    def scale (self, row1: int, operator: int, operand: tuple):
        if operator != 0 and operator != 1:
            print("Scaling supports multilpicaton and division. Type 0 for division, 1 for multiplication.")
            return
        numerator = operand[0]
        denominator = operand[1]
        if operand[1] == 0:
            print("Division by zero is not allowed")
            return
        if operator == 0:
            if operand[0] == 0:
                print("Entries cannot be divided by zero")
                return
            #flipping the numerator and denominator of the operand so that I can avoid division
            operand = (denominator, numerator)
            #print (operand) #scaffolding
        #scaling the row
        for i in range (len(self.entries[row1-1])):
            #multiplying by the operand(or one over the operand for division - switched above)
            placeholder_numerator = self.entries[row1-1][i][0] * operand[0]
            placeholder_denominator = self.entries[row1-1][i][1] * operand[1]
            #reducing the resulting fraction
            new_tuple = reduce_frac(placeholder_numerator, placeholder_denominator)
            self.entries[row1-1][i] = (new_tuple[0], new_tuple[1])
        return (self)
    

    def swap (self, row1: int,row2: int):
        temp = self.entries[row1 - 1]
        self.entries[row1 - 1] = self.entries[row2 - 1]
        self.entries[row2 - 1] = temp
        return (self)
    
    def add(self, change_row: int, scale_row: int, multiplier: tuple )-> list:
    #for i in self.entries[change_row - 1]:
        for i in range (len(self.entries[change_row-1])):
            temp_tuple = fraction_multiplier(self.entries[scale_row-1][i],multiplier)
            #print(temp_tuple[1])
            #print(self.entries[change_row - 1][i][1])


            #if the denominators of the two tuples are the same, add the tuples, reduce, and edit the matrix
            if temp_tuple[1] == self.entries[change_row - 1][i][1]:
                #print("hi")
                self.entries[change_row-1][i] = reduce_frac(temp_tuple[0] + self.entries[change_row - 1][i][0],self.entries[change_row - 1][i][1])
                #print(self.entries)
            else: #need to have a common denominator in order to add 
            #if the denominators of the two are different, cross-multiply and reduce

                large_den_temp1 = (self.entries[change_row - 1][i][0] * temp_tuple[1],self.entries[change_row - 1][i][1] * temp_tuple[1])
                #print(large_den_temp1)
                large_den_temp2 = (temp_tuple[0] * self.entries[change_row - 1][i][1],temp_tuple[1] * self.entries[change_row - 1][i][1])
                #print(large_den_temp2)
                new_tuple_temp = reduce_frac(large_den_temp1[0] + large_den_temp2[0],large_den_temp1[1])
                self.entries[change_row-1][i] = new_tuple_temp
        return self

    


def main():
     """Main program driver for row reduction calculator"""
     
     #matrix = input("Please input the matrix:")
     #testy_test = Matrix((3,3),(0,0),[[(1,1),(1,2),(1,10)],[(0,1),(-7,1),(-4,1)],[(3,1),(1,1),(-1,1)]])
     #print(testy_test.row_scale(1,1,(2,1)))
     #print(testy_test.row_swap(1,3))
     #print([testy_test])
     
     #testing with an integer matrix from a homework assignment. This works.
     """ hw_test_m = Matrix((3,1),(0,0),[[(1,1),(0,1),(2,1)],[(0,1),(1,1),(3,1)],[(-2,1),(2,1),(3,1)]])
     print(hw_test_m)
     print("")
     step1 = hw_test_m.add(3,1,(2,1))
     print(step1)
     print("")
     step2 = step1.add(3,2,(-2,1))
     print(step2)
     print("")
     step3 = step2.add(2,3,(-3,1))
     print(step3)
     print("")
     step4 = step3.add(1,3,(-2,1))
     print(step4) """

     #testing with a matrix from a homework that involves fractions with different denominators
     #this now works
     hw_fraction_test = Matrix((4,5),(0,0),[[(3,1),(9,1),(2,1),(1,1),(3,1)],[(0,1),(3,1),(-4,1),(1,1),(-1,1)],[(1,1),(2,1),(3,1),(0,1),(4,1)],[(3,1),(0,1),(5,1),(-2,1),(-1,1)]])
     print(hw_fraction_test)
     print("")
     step1 = hw_fraction_test.swap(1,3)
     print(step1)
     print("")
     step2 = step1.swap(2,4)
     print(step2)
     print("")
     step3 = step2.swap(2,3)
     print(step3)
     print("")
     step4 = step3.add(3,1,(-3,1))
     print(step4)
     print("")
     step5 = step4.scale(3,1,(1,2))
     print(step5)
     print("")
     step6 = step5.add(2,1,(-3,1))
     print(step6)
     print("")
     step7 = step6.add(4,3,(1,1))
     print(step7)
     print("")
     step8 = step7.add(3,2,(1,1))
     print(step8)
     print("")
     step9 = step8.scale(3,1,(-2,3))
     print(step9)
     print("")
     step10 = step9.add(4,3,(1,1)) 
     print(step10)



if __name__ == "__main__":
    main()

File: Matrix.js
 // Important! 
// To use node for testing the code, you must first type:
//  \. "$HOME/.nvm/nvm.sh"
//After that, typing 'node test2.js' will work
 
 // Carmen's gcd function:
//helper greatest common divisor function
function gcd(a, b) {
    a = Math.abs(a);
    b = Math.abs(b);
    while (b !== 0) {
        const temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}


// a helper function to reduce a fraction
function reduce_frac (x,y) { // x is numerator, y is denominator
    let gCD = gcd(x,y) 
    x = Math.floor(x/gCD);
    y = Math.floor(y/gCD);
    
    return [x,y] 
}

//a helper function to multiply (and reduce) fractions
function fraction_multiplier(no_1,no_2) {///multiplies two arrays with numerator in position 0 and denominator in pos. 1
    const numerator = no_1[0] * no_2[0];
    const denominator = no_1[1] * no_2[1];
    let solution = reduce_frac(numerator, denominator);
    return solution
}
 
 
 /* A matrix has a size mxn, state (which version/step it is in the reduction process) and entries represented as arrays of length 2, 
    with numerator as the first entry and denominator as the second.
    Entries are an array with an array of arrays(entries) for each row.
    Usage example:
        matrix1 = Matrix([3,2], [0,1], [[[2,1],[4,1]],
                                        [[0,1],[-7,1]],
                                        [[3,1],[1,1]]])
    */

class Matrix {
    constructor(size, state = 0, entries) {
        this.size = size;
        this.state = state;
        this.entries = entries;
    }

    //im adding init entries to help me
    init_entries() {
        let rows = this.size[0];
        let cols = this.size[1];
        let temp = [];

        for (let i = 0; i < rows; i++) {
            let row = [];
            for (let j = 0; j < cols; j++) {
                row.push([0,1]); //initializing all entries to nuthin
            }
            temp.push(row);
        }   
        this.entries = temp;//no return here, just setting space for entries
    }


    print() {
        console.log("Matrix Values:");
        this.entries.forEach(row => {
        console.log(row.map(entry => `[${entry[0]},${entry[1]}]`).join(' '));
        });
    }
    

    //also need an add entry function
    add_value(row, col, value) { //value is an array [numerator, denominator]
        this.entries[row-1][col-1] = value; 
    }

    //way to grab an entry
    get_entry(row, col) {
        return this.entries[row-1][col-1]; 
    }   

    swap(row1,row2) { //this method works. 7.30.25
        let temp = this.entries[row1-1]
        //console.log(temp);
        this.entries[row1-1] = this.entries[row2-1];  
        //console.log(this.entries[row1-1]);
        this.entries[row2-1] = temp; 
        return this.entries;

    }

    //operator: 0 = division, 1 = multiplication. operand: [numerator,denominator]
    // operand is entered as an array of two entries, the first corresponding to the numerator and the second the denominator
    scale (row1, operator, operand) {
    
        //console.log(row1); // scaffolding
        //console.log(operator); // scaffolding
        //console.log(operand[0]); // scaffolding
        let numerator = operand[0];
        let denominator = operand[1];
        if (operand[1] == 0) {
            return;
        } // to prevent division by zero. probably can be removed from final verson b/c check will occur in main
        
        if (operator == 0) { //if the user wishes to divide a row by a number
            if (operand[0] == 0) {
                return;
            } //this prevention against division by zero can probably be removed, see similar comment above

            //flipping the numerator and denominator of the operand so that I can avoid division
            operand = [denominator, numerator]
            //console.log(operand) //scaffolding
        }

         //scaling the row
        //console.log(this.entries[row1-1].length); //scaffolding
        let row_len = this.entries[row1-1].length;
        //console.log("Row length is:"); //scaffolding
        //console.log(row_len); // scaffolding
        for (let i = 0; i < row_len; i++) { 
            //multiplying by the operand(or one over the operand for division - switched above)
            let placeholder_numerator = this.entries[row1-1][i][0] * operand[0];
            //console.log(placeholder_numerator); // scaffolding
            let placeholder_denominator = this.entries[row1-1][i][1] * operand[1];
            //console.log(placeholder_denominator); //scaffolding

            //reducing the resulting fraction
            let new_tuple = reduce_frac(placeholder_numerator, placeholder_denominator); //this is getting rid of the denominator!
            //console.log("New tuple"); //scaffolding
            //console.log(new_tuple); //scaffolding
            this.entries[row1-1][i] = [new_tuple[0], new_tuple[1]];
        }
        return this.entries;
    }


    //note: multiplier parameter must be passed in as an array. Ex: 1/2 is entered as [1,2]; -8 is entered as [-8,1]
    // change row is entered as an int, as is scale_row
    add(change_row, scale_row, multiplier) {

        let row_len = this.entries[change_row-1].length;
        for (let i = 0; i < row_len; i++) { 
    
            let temp_tuple = fraction_multiplier(this.entries[scale_row-1][i],multiplier); //this scales the appropriate row before adding
            //console.log(temp_tuple); // scaffolding
            //console.log(this.entries[scale_row - 1][i]); //scaffolding

            //if the denominators of the two arrays are the same, add the arrays, reduce,  and edit the matrix
            if (temp_tuple[1] == this.entries[change_row - 1][i][1]) {
                //console.log("same"); //scaffolding
                this.entries[change_row-1][i] = reduce_frac(temp_tuple[0] + this.entries[change_row - 1][i][0],this.entries[change_row - 1][i][1]);
            }
            
        else { //need to have a common denominator in order to add 
            //if the denominators of the two are different, cross-multiply and reduce

            //console.log('different'); //scaffolding
            let large_den_temp1 = [this.entries[change_row - 1][i][0] * temp_tuple[1],this.entries[change_row - 1][i][1] * temp_tuple[1]];
            //console.log(large_den_temp1); //scaffolding
            let large_den_temp2 = [temp_tuple[0] * this.entries[change_row - 1][i][1],temp_tuple[1] * this.entries[change_row - 1][i][1]];
            //console.log(large_den_temp2); //scaffolding
            let new_tuple_temp = reduce_frac(large_den_temp1[0] + large_den_temp2[0],large_den_temp1[1]);
            //console.log(new_tuple_temp); // scaffolding
            this.entries[change_row-1][i] = new_tuple_temp;
        }
    }

    return this.entries;
}

}



//let test_matrix = new Matrix ([3,2], [0,1], [[[2,1],[4,1]],[[0,1],[-7,1]],[[3,1],[1,2]]])
//let test_matrix2 = new Matrix ([3,2], [0,1], [[[2,1],[4,1]],[[0,1],[-7,1]],[[3,1],[-15,2]]])

//console.log(test_matrix)

//console.log("Testing...");
//console.log("Original matrix");
//console.log(test_matrix.entries);

// testing swap
//console.log("Swapped Rows 2 and 3")
//console.log(test_matrix.swap(2,3))

//testing scale
//console.log("Multiplied row 2 by 2");
//console.log(test_matrix.scale(2,1,[2,1]));
//console.log("Divided row 3 by 2");
//console.log(test_matrix.scale(3,0,[2,1]))
   
//testing add
//fraction test with different denominators
//let fraction_test = new Matrix([4,5],[0,1],[[[1,1],[2,1],[3,1],[0,1],[4,1]],[[0,1],[3,1],[-7,1],[1,1],[-9,1]],[[0,1],[0,1],[6,1],[0,1],[31,3]],[[0,1],[0,1],[-6,1],[0,1],[-15,2]]]);


let fraction_test2 = new Matrix ([2,2], [0,1], [[[1,1],[31,3]],[[0,1],[-15,2]]]);
console.log(fraction_test2.add(1,2,[62,45])); //should return [ [ [ 1, 1 ], [ 0, 1 ] ], [ [ 0, 1 ], [ -15, 2 ] ] ]
//fraction test 2 works

/*
//this is a matrix from a homework assignment. 'add' method works on this. Note: no fractions.
//final matrix is [[[1,1],[0,1],[0,1]],[[0,1],[1,1],[0,1]],[[0,1],[0,1],[1,1]]]
let hw_test_m = new Matrix ([3,3],[0,1],[[[1,1],[0,1],[2,1]],[[0,1],[1,1],[3,1]],[[-2,1],[2,1],[3,1]]]);
console.log("Original matrix:");
console.log(hw_test_m.entries);
let step1 = hw_test_m.add(3,1,[2,1]);
console.log("Step 1");
console.log(step1);
let step2 = hw_test_m.add(3,2,[-2,1]);
console.log("Step 2");
console.log(step2);
let step3 = hw_test_m.add(2,3,[-3,1]);
console.log("Step 3");
console.log(step3);
let step4 = hw_test_m.add(1,3,[-2,1]);
console.log("Row reduced:");
console.log(step4);
*/